### PR TITLE
gh-115765: Don't use deprecated AC_EGREP_* macros in configure.ac

### PR DESCRIPTION
**Pull request by erlend-aasland**
See Autoconf docs:

- https://www.gnu.org/software/autoconf/manual/autoconf-2.72/html_node/Running-the-Preprocessor.html#index-AC_005fEGREP_005fHEADER-1

* Issue: gh-115765

### CODE REVIEW

**Comment by erlend-aasland** [comment]
@mhsmith, would you like to review this?

**Comment by erlend-aasland** [comment]
@corona10, feel free to take a look at this, if you have the time.

File: configure.ac [hunk 1]
```diff
@@ -73,6 +73,17 @@ AC_DEFUN([PY_CHECK_FUNC],
...
 ])
 
+dnl PY_CHECK_CPP(CPP_CHECK, [LIST-OF-INCLUDES], [ACTION-IF-FOUND], [ACTION-IF-NOT-FOUND])
+AC_DEFUN([PY_CHECK_CPP],
+         [AC_PREPROC_IFELSE([AC_LANG_PROGRAM([
+m4_foreach_w([incl], [$2], [@%:@include <incl> m4_newline])
+@%:@if $1
+/* pass */
+#else
+#error "not defined"
```
Comment (mhsmith) [inline configure.ac]: This function is more general than just checking for a definition.
```suggestion
dnl PY_CHECK_CPP(CPP_CHECK, [LIST-OF-INCLUDES], [ACTION-IF-TRUE], [ACTION-IF-FALSE])
AC_DEFUN([PY_CHECK_CPP],
         [AC_PREPROC_IFELSE([AC_LANG_PROGRAM([
m4_foreach_w([incl], [$2], [@%:@include <incl> m4_newline])
@%:@if $1
/* pass */
#else
#error "check returned false"
```

File: configure.ac [hunk 2]
```diff
@@ -2829,7 +2840,7 @@ AC_CHECK_HEADERS(
 
 # checks for typedefs
 AC_CACHE_CHECK([for clock_t in time.h], [ac_cv_clock_t_time_h], [
-  AC_EGREP_HEADER([clock_t], [time.h], [ac_cv_clock_t_time_h=yes], [ac_cv_clock_t_time_h=no])
+  AX_CHECK_TYPEDEF([clock_t], [time.h], [ac_cv_clock_t_time_h=yes], [ac_cv_clock_t_time_h=no])
```
Comment (mhsmith) [inline configure.ac]: I'm still pretty new to autoconf, but it looks to me like the [definition of AX_CHECK_TYPEDEF](https://git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=blob_plain;f=m4/ax_check_typedef.m4) ignores its third and fourth arguments and replaces them with AC_MSG_RESULT.

[This project](https://github.com/NumCosmo/NumCosmo/commit/203ea7469bc45ffa19e693c8149edbc5603cba22) may have hit the same problem, and responded by replacing AX_CHECK_TYPEDEF with AC_CHECK_TYPES.

File: configure.ac (hunk 1 above)
Comment (erlend-aasland) [reply configure.ac]: Thanks, good observation!

File: configure.ac (hunk 2 above)
Comment (erlend-aasland) [reply configure.ac]: Hm, I did not study the generated output good enough. The cv variable now has a slightly different format. It seems to me should apply this patch (diff against PR):

```diff
diff --git a/configure.ac b/configure.ac
index 6b551b699d..c18343a11b 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2838,11 +2838,11 @@ AC_CHECK_HEADERS(
 ])
 
 # checks for typedefs
-AC_CACHE_CHECK([for clock_t in time.h], [ac_cv_clock_t_time_h], [
-  AX_CHECK_TYPEDEF([clock_t], [time.h], [ac_cv_clock_t_time_h=yes], [ac_cv_clock_t_time_h=no])
-])
+AC_CACHE_CHECK([for clock_t in time.h],
+               [ac_cv_type_clock_t_time_h],
+               [AX_CHECK_TYPEDEF([clock_t], [time.h])])
 dnl checks for "no"
-AS_VAR_IF([ac_cv_clock_t_time_h], [no], [
+AS_VAR_IF([ac_cv_type_clock_t_time_h], [no], [
   AC_DEFINE([clock_t], [long],
             [Define to 'long' if <time.h> doesn't define.])
 ])

```

Comment (erlend-aasland) [reply configure.ac]: > [This project](https://github.com/NumCosmo/NumCosmo/commit/203ea7469bc45ffa19e693c8149edbc5603cba22) may have hit the same problem, and responded by replacing AX_CHECK_TYPEDEF with AC_CHECK_TYPES.

That's also an option, but `AC_CHECK_TYPES` introduces a new definition to pyconfig.h. OTOH, `AX_CHECK_TYPEDEF` introduces a new macro to aclocal.m4.

Comment (erlend-aasland) [reply configure.ac]: Another nice thing with `AC_CHECK_TYPES`, is that it has caching automatically built in.

Comment (erlend-aasland) [reply configure.ac]: Diff for `AC_CHECK_TYPES`:

```diff
diff --git a/configure.ac b/configure.ac
index 6b551b699d..3e1ea3ff05 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2837,15 +2837,13 @@ AC_CHECK_HEADERS(
 #endif
 ])
 
-# checks for typedefs
-AC_CACHE_CHECK([for clock_t in time.h], [ac_cv_clock_t_time_h], [
-  AX_CHECK_TYPEDEF([clock_t], [time.h], [ac_cv_clock_t_time_h=yes], [ac_cv_clock_t_time_h=no])
-])
-dnl checks for "no"
-AS_VAR_IF([ac_cv_clock_t_time_h], [no], [
-  AC_DEFINE([clock_t], [long],
-            [Define to 'long' if <time.h> doesn't define.])
-])
+# Check for clock_t in time.h.
+AC_CHECK_TYPES([clock_t],
+               [],
+               [AS_VAR_IF([ac_cv_type_clock_t], [no],
+                          [AC_DEFINE([clock_t], [long],
+                                     ["Define to 'long' if <time.h> doesn't define."])])],
+               [@%:@include time.h])
 

```

**Comment by erlend-aasland** [comment]
@mhsmith, thanks for the review; I addressed your remarks in 9c3596d57520e0ca913119757a61cd4fad794839. PTAL.

File: pyconfig.h.in [hunk 3]
```diff
@@ -1938,7 +1941,7 @@
 /* Define on FreeBSD to activate all library features */
 #undef __BSD_VISIBLE
 
-/* Define to 'long' if <time.h> doesn't define. */
+/* Define to 'long' if <time.h> does not define clock_t. */
```
Comment (erlend-aasland) [inline pyconfig.h.in:1944]: Because three single quotes on a line messes up syntax highlighting in Vim 💥

File: configure.ac (hunk 2 above)
Comment (mhsmith) [reply configure.ac]: Is the `AS_VAR_IF` necessary? It's already in the "not found" branch.

Comment (erlend-aasland) [reply configure.ac]: Hah, true 😆 I guess I was in a hurry earlier today; looks like my last commit broke CI quite bad.

File: configure [hunk 4]
```diff
@@ -16097,20 +16083,28 @@ printf %s "checking for _POSIX_THREADS in unistd.h... " >&6; }
...
+#error "check failed"
 #endif
 
+int
+main (void)
+{
+
+  ;
+  return 0;
+}
```
Comment (mhsmith) [inline configure]: Could this noise be reduced by using AC_LANG_SOURCE rather than AC_LANG_PROGRAM?

Comment (erlend-aasland) [reply configure]: Definitely! Good catch.

**Comment by erlend-aasland** [comment]
Looking at my local configure run, I see that the `AC_CHECK_TYPES` check for `clock_t` in `time.h` fails, but it succeeds for `AX_CHECK_TYPEDEF`. The checks performed are similar, but not equal.

The former does:

```c
#include <time.h>
int
main (void)
{
if (sizeof(clock_t)) 
      return 0;
  ;
  return 0;
}
```

The latter does:

```c
#include <time.h>
int
main (void)
{
int x = sizeof(clock_t); x = x;
  ;
  return 0;
}
```

**Comment by erlend-aasland** [comment]
Found the culprit; typo on my end! ✅

**Comment by erlend-aasland** [comment]
Now, there's one more thing we should add before landing this: support configure cache.

**Comment by erlend-aasland** [comment]
> Now, there's one more thing we should add before landing this: support configure cache.

I'm not sure we want to automatically generate the cache variable name in this macro. The first argument would need heavy string modifications to end up as a usable identifier. Perhaps it is best to introduce a new param for the cache variable name.

**Comment by erlend-aasland** [comment]
> Alternatively, all but one usage of PY_CHECK_CPP could be replaced with [AX_CHECK_DEFINE](https://www.gnu.org/software/autoconf-archive/ax_check_define.html), which has caching already built in.

That's a good suggestion; I'll have a look in a day or two.

**Comment by erlend-aasland** [comment]
> Alternatively, all but one usage of PY_CHECK_CPP could be replaced with [AX_CHECK_DEFINE](https://www.gnu.org/software/autoconf-archive/ax_check_define.html), which has caching already built in.

For the glibc check; do we really need to worry about glibc < 2.1?

**Comment by mhsmith** [comment]
I agree, so #116016 looks like the way to go.

**Comment by erlend-aasland** [comment]
Superseded by #116016

**Pull request closed without merging**